### PR TITLE
Infinite stream workarounds (using kind-of affine streams underneath)

### PR DESCRIPTION
**Pull request by Divesh-Otwani**
## Overview

This PR replaces #31 and adds workarounds for most of the infinite streaming API from the original library internally using an unfolded affine stream data type mentioned in #31. We defer `never` for another day, leaving issue #32.  The API for `repeats` and `repeatsM` are the same as the `replicate` functions and hence there is no `repeatsZip`, for example. (More detail if needed: taking, until-ing and zipping with repeats is the same as taking, until-ing and zipping variants of `replicates`.)

As before, the rationale for the chosen API (with edits for the markdown and such) is in the readme (cut and paste below for easy reading).

## Approach

We internally use the following data type from the discussion in #31:

```haskell
data AffineStream f m r where
  AffineStream :: x #-> (x #-> m (Either (f x) r)) -> (x #-> m r) -> AffineStream f m r
```

and implement the API via one-liners that use

```haskell
take :: forall f m r. (Control.Monad m, Control.Functor f) =>
  Int -> AffineStream f m r #-> Stream f m r
untilM :: forall a m r. Control.Monad m =>
  (a -> m Bool) -> AffineStream (Of a) m r #-> Stream (Of a) m r
until :: forall a m r. Control.Monad m =>
  (a -> Bool) -> AffineStream (Of a) m r #-> Stream (Of a) m r
zip :: forall a x m r1 r2. Control.Monad m =>
  Stream (Of x) m r1 #-> AffineStream (Of a) m r2 #-> Stream (Of (x,a)) m (r1,r2)
```

## Cut and paste readme:

```markdown
### Replacing infinite streams like `stdinLn`

We can't have functions like `stdinLn` that produce infinite streams.
Any function that linearly consumes a `Stream (Of a) m r` where the `m`
is a control monad, would have to consume the entire stream. Hence,
for some `f :: Stream (Of a) m r #-> B`, `f stdinLn` would never terminate.
Hence, we need workarounds for the original API.

The infinite stream API in the original library consisted of
`repeats`, `repeatsM`, `stdinLn`, `readLn`, `cycle`, `enumFrom`, `enumFromThen`.

The size-delimited replacements of `repeats` and `repeatsM` are just `replicates` and
`replicatesM` from the `Streaming` module.

The rest of the API are constructors that produce an infinite streams `Of` elements
possibly from some simple arguments. From combing the examples in the Haddock and linked
to by the haddock, three common use cases emerge:

 1. Taking some finite amount:

 `stdoutLn$ doSomething $ take 3 $ stdinLn`

 2. Taking until a condition is met:

 `highLowGame n = void $ S.break (== n) (readLn :: Stream (Of Int) m ())`

 3. Zipping with a finite stream:

  `doSomething $ zip (each' ["first name: ", "last name: ", "nickname: "]) stdinLn`

 4. Some combination of the above.

We can avoid/replace these cases by having these variants of infinite
streams like `stdinLn`:

`stdinLnN :: Int -> Stream (Of Text) IO ()`
`stdinLnUntil :: (Text -> Bool) -> Stream (Of Text) IO ()`
`stdinLnUntilM :: (Text -> IO Bool) -> Stream (Of Text) IO ()`
`stdinLnZip :: (Stream (Of a) m ()) #-> Stream (Of (a, Text)) IO ()`

This approach is adopted for the rest of the API and the postfixes
`N`, `Until`, `UntilM` and `Zip` are followed. We don't need `Until*` functions
for `replicate`, `cycle`, `enumFrom` and `enumFromThen` since these streams are simple
enough that a test function doesn't make sense; we can predict each element so there's no
cause to test until a condition is met.

```

### CODE REVIEW

File: src/Streaming/Internal/Produce.hs [hunk 1]
```diff
@@ -156,3 +186,297 @@ untilRight mEither = Effect loop
...
 
+
+-- # The \"Affine\" 'Stream'
+-------------------------------------------------------------------------------
+
+-- | An *affine stream is represented with a state of type @x@,
+-- a possibly terminating step function of type @(x #-> m (Either (f x) r))@,
+-- and a stop-short function @(x #-> m r)@.
+--
+-- This mirrors the inverse unfold of a normal stream:
```
Comment (aspiwack) [inline src/Streaming/Internal/Produce.hs:197]: inverse unfold?

**Review by aspiwack — APPROVED** [review]
You could have just pushed a commit on the other PR. I don't think this was deserving a whole new PR.

File: src/Streaming/Internal/Produce.hs (hunk 1 above)
Comment (aspiwack) [reply src/Streaming/Internal/Produce.hs:197]: This is still here though, and still doesn't make much sense, can you explain to me what you meant?

Comment (Divesh-Otwani) [reply src/Streaming/Internal/Produce.hs:197]: I didn't push the local changes I made. I was going to re-write it to just `unfold`; I made the change in master just now. I originally meant the inverse of the `unfold` function.

Comment (aspiwack) [reply src/Streaming/Internal/Produce.hs:197]: Ah, I see. It's a matter of point of view, I suppose. I guess that this is saying that `unfold` is an isomorphism. I would have just said that it mirrors the `unfold` function myself.

**Pull request merged**